### PR TITLE
Horizontal scroll; replace pending input by suggestion

### PR DESCRIPTION
**Pull request by JaxGit**
This is intended to achieve the following targets: 
1. be able to replace current pending input by suggestions like autocompletions; 
2. be able to scroll horizontally like iOS Instagram "Direct" -> "New Message" recipient input component.

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -227,7 +253,17 @@ class TagInput<T> extends React.PureComponent<OptionalProps, Props<T>, State> {
 
     if (results && results.length > 0) {
       this.setState({ text: '' });
-      this.props.onChange([...new Set([...value, ...results])]);
+      this.onChangeTags([...new Set([...value, ...results])]);
+    }
+  }
+
+  // Public useful
+  addCustomTag = (newTag) => {
```
Comment (Ashoat) [inline index.js:263]: Does this typecheck without a specified type? Did you run Flow before making the PR?

File: Example/src/TagInputExample.js [hunk 2]
```diff
@@ -16,18 +19,40 @@ export default class TagInputExample extends Component {
...
+  onChangeHorizontalTags = (horizontalTags) => {
+    this.setState({
+      horizontalTags,
+    });
+  };
+
   labelExtractor = (tag) => tag;
 
+  onParsePendingInput = () => {
+    this.refs.horizontalTagInput.parseTags();
```
Comment (Ashoat) [inline Example/src/TagInputExample.js:31]: I'm not a fan of having `TagInput` hold on to a piece of state - in this case, the `text` state - that its parent needs to manipulate.

In situations where multiple React components need to manipulate state, that state should be ["lifted up to their closest common ancestor"](https://reactjs.org/docs/lifting-state-up.html). In other words - if the parent component and child component both need to maintain a piece of state, then the parent should be holding on to that state.

In general, refs are a hack around this, and should be used sparingly and only when necessary.

More discussion in my higher-level review comment.

File: Example/src/TagInputExample.js [hunk 3]
```diff
@@ -39,6 +64,39 @@ export default class TagInputExample extends Component {
             maxHeight={75}
           />
         </View>
+
+        <Text style={{marginVertical: 10}}>Horizontal Scroll</Text>
+        <View style={{marginBottom: 10, flexDirection: 'row', alignItems: 'center', backgroundColor: 'lightblue'}}>
+          <Text>To: </Text>
+          <TagInput
+            ref='horizontalTagInput'
```
Comment (Ashoat) [inline Example/src/TagInputExample.js:76]: Notwithstanding my comment above, string values from the ref parameter are discouraged in favor of using ref functions, which typecheck much better.

File: Example/android/app/build.gradle [hunk 4]
```diff
@@ -96,7 +96,7 @@ android {
     defaultConfig {
         applicationId "com.example"
         minSdkVersion 16
-        targetSdkVersion 22
+        targetSdkVersion 24
```
Comment (Ashoat) [inline Example/android/app/build.gradle:99]: Why is this necessary?

File: index.js [hunk 5]
```diff
@@ -227,7 +253,17 @@ class TagInput<T> extends React.PureComponent<OptionalProps, Props<T>, State> {
 
     if (results && results.length > 0) {
       this.setState({ text: '' });
-      this.props.onChange([...new Set([...value, ...results])]);
+      this.onChangeTags([...new Set([...value, ...results])]);
+    }
+  }
+
+  // Public useful
```
Comment (Ashoat) [inline index.js:262]: This comment is not helpful, please elaborate

File: Example/src/TagInputExample.js [hunk 6]
```diff
@@ -16,18 +19,40 @@ export default class TagInputExample extends Component {
...
+  };
+
   labelExtractor = (tag) => tag;
 
+  onParsePendingInput = () => {
+    this.refs.horizontalTagInput.parseTags();
+  }
+
+  onaddCustomTag = () => {
+    this.refs.horizontalTagInput.addCustomTag('Suggested Tag');
```
Comment (Ashoat) [inline Example/src/TagInputExample.js:35]: Same comment as above - this shouldn't be accessed via a ref. It would not be necessary if the `text` state were lifted.

File: Example/src/TagInputExample.js [hunk 7]
```diff
@@ -16,18 +19,40 @@ export default class TagInputExample extends Component {
...
+    });
+  };
+
   labelExtractor = (tag) => tag;
 
+  onParsePendingInput = () => {
+    this.refs.horizontalTagInput.parseTags();
+  }
+
+  onaddCustomTag = () => {
```
Comment (Ashoat) [inline Example/src/TagInputExample.js:34]: Capitalize the "a" in "Add"

File: README.md [hunk 8]
```diff
@@ -33,3 +33,11 @@ import TagInput from 'react-native-tag-input';
...
 | onHeightChange | Callback that gets passed the new component height when it changes |
 | parseOnBlur | Whether to treat a blur event as a separator entry (iOS-only) |
+| parseOnSubmit | Whether to treat submit button press as a separator entry |
+| onChangeText | A handler to be called when input text has changed |
+| scrollHorizontal | Whether the wrapper scrollView should scroll horizontally |
+| scrollViewProps | Any misc. ScrollView props (showsHorizontalScrollIndicator, etc.) |
+
+| Public Useful Method | Description |
+-----------------------|-----------------
+| addCustomTag | Replace pending input with suggested custom tag |
```
Comment (Ashoat) [inline README.md:44]: You forgot to document "parseTags", which you are using from the example. That said, I don't think we should have any "public useful methods", and should just raise the `text` state.

**Review by Ashoat — CHANGES_REQUESTED** [review]
First of all, thank you for the PR.

Some high-level comments:

1. If you notice yourself making a PR that is just two different things put together, please take the time to separate your pieces out. Can you open a separate PR for the horizontal scroll piece?
2. In this PR, you find yourself needing to make decisions about manipulating the `text` state in TagInput from the parent component. This is a good sign that the [state should be lifted](https://reactjs.org/docs/lifting-state-up.html). I found myself needing the same functionality, so I made a proposal [here](https://github.com/jwohlfert23/react-native-tag-input/issues/39).
3. Regarding the horizontal scroll - this seems like something we should support. I'd like to understand the need for the `shouldScrollIfHorizontal` behavior, and why this doesn't apply in a vertical scenario. And as noted above, please submit horizontal scroll stuff as a separate PR!

**Comment by JaxGit** [comment]
@Ashoat thank you for your response, I agree that independent features should better be separated into different PRs, I was lazy and also considering the new example finally needs both features. I'll try to answer some minor questions and separate these in a few days.

File: Example/android/app/build.gradle (hunk 4 above)
Comment (JaxGit) [reply Example/android/app/build.gradle:99]: Got build error of "Package com.example new target SDK 22 doesn't support runtime permissions but the old target SDK 24 does"... 
After deleting app/clearing cache "targetSdkVersion 22" is actually working. 👍

File: Example/src/TagInputExample.js (hunk 2 above)
Comment (JaxGit) [reply Example/src/TagInputExample.js:31]: 👍

File: README.md (hunk 8 above)
Comment (JaxGit) [reply README.md:44]: 👍

**Comment by Ashoat** [comment]
I went ahead and created a PR for lifting the text state: #43

**Comment by JaxGit** [comment]
@Ashoat I extracted horizontal scroll and create a PR: #44 
This can be closed :)
